onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -0,0 +1,401 @@
+        outputOperatorDesc.OutputTensor = &namedTransposedOutputTensorDesc;
+        const DML_OPERATOR_DESC outputDesc {DML_OPERATOR_ACTIVATION_LINEAR, &outputOperatorDesc};
+
+
+        MLOperatorGraphDesc operatorGraphDesc = {};
+        operatorGraphDesc.nodeCount = 10;
+        std::array<const DML_OPERATOR_DESC*, 10> opDescs{&xWeightDesc, &querySlicedDesc,     &keySlicedDesc,     &valueSlicedDesc, &attentionScoreDesc, 
+                                                         &softmaxDesc, &attentionWeightDesc, &castMaskIndexDesc, &maskDesc,        &outputDesc};
+        operatorGraphDesc.nodesAsOpDesc = opDescs.data();
+
+        // set input edges
+        std::pair<uint32_t, uint32_t> nodeToNodeInputIndex[4] {{0, 0}, {0, 1}, {0, 2}, {7, 0}};
+        std::array<DML_INPUT_GRAPH_EDGE_DESC, 4> inputEdges;
+        for (uint32_t inputIndex = 0; inputIndex < 4; inputIndex++)
+        {
+            DML_INPUT_GRAPH_EDGE_DESC inputEdge = {};
+            inputEdge.GraphInputIndex = inputIndex;
+            inputEdge.ToNodeIndex = nodeToNodeInputIndex[inputIndex].first;
+            inputEdge.ToNodeInputIndex = nodeToNodeInputIndex[inputIndex].second;
+            inputEdges[inputIndex] = inputEdge;
+        }
+        operatorGraphDesc.inputEdgeCount = gsl::narrow_cast<uint32_t>(inputEdges.size());
+        operatorGraphDesc.inputEdges = inputEdges.data();
+
+        // set intermediate edges
+        std::vector<DML_INTERMEDIATE_GRAPH_EDGE_DESC> intermediateEdges;
+        
+        DML_INTERMEDIATE_GRAPH_EDGE_DESC gemmToQuerySliceEdge = {};
+        gemmToQuerySliceEdge.FromNodeIndex = 0;
+        gemmToQuerySliceEdge.FromNodeOutputIndex = 0;
+        gemmToQuerySliceEdge.ToNodeIndex = 1;
+        gemmToQuerySliceEdge.ToNodeInputIndex = 0;
+        intermediateEdges.push_back(gemmToQuerySliceEdge);
+
+        DML_INTERMEDIATE_GRAPH_EDGE_DESC gemmToKeySliceEdge = {};
+        gemmToKeySliceEdge.FromNodeIndex = 0;
+        gemmToKeySliceEdge.FromNodeOutputIndex = 0;
+        gemmToKeySliceEdge.ToNodeIndex = 2;
+        gemmToKeySliceEdge.ToNodeInputIndex = 0;
+        intermediateEdges.push_back(gemmToKeySliceEdge);
+
+        DML_INTERMEDIATE_GRAPH_EDGE_DESC gemmToValueSliceEdge = {};
+        gemmToValueSliceEdge.FromNodeIndex = 0;
+        gemmToValueSliceEdge.FromNodeOutputIndex = 0;
+        gemmToValueSliceEdge.ToNodeIndex = 3;
+        gemmToValueSliceEdge.ToNodeInputIndex = 0;
+        intermediateEdges.push_back(gemmToValueSliceEdge);
+
+        DML_INTERMEDIATE_GRAPH_EDGE_DESC querySliceToGemm = {};
+        querySliceToGemm.FromNodeIndex = 1;
+        querySliceToGemm.FromNodeOutputIndex = 0;
+        querySliceToGemm.ToNodeIndex = 4;
+        querySliceToGemm.ToNodeInputIndex = 0;
+        intermediateEdges.push_back(querySliceToGemm);
+
+        DML_INTERMEDIATE_GRAPH_EDGE_DESC keySliceToGemm = {};
+        keySliceToGemm.FromNodeIndex = 2;
+        keySliceToGemm.FromNodeOutputIndex = 0;
+        keySliceToGemm.ToNodeIndex = 4;
```

Comment:
>4 [](http://example.com/codeflow?start=37&length=1)

The edge desc names are helpful, and the ASCII diagram too, but I still pity anyone who has to add new ops or reorder ones. Rather than numbers which are brittle to being moved around, I propose enums like we do in BatchNormalization. If you add an enum list right next to ` std::array<const DML_OPERATOR_DESC*, 10> opDescs`, then the relationship should be clear 1:1.
```
enum NodeIndex
{
    xWeight,
    querySlice,
    ...
}
...

        DML_INTERMEDIATE_GRAPH_EDGE_DESC keySliceToGemm = {};
        keySliceToGemm.FromNodeIndex = NodeIndex::keySlice;
        keySliceToGemm.FromNodeOutputIndex = 0;
        keySliceToGemm.ToNodeIndex = NodeIndex::maskDesc??; // I don't know which gemm because there are 3 GEMM's.
        keySliceToGemm.ToNodeInputIndex = 1;
        intermediateEdges.push_back(keySliceToGemm);
```

Reply:
I agree it will make the readability and modification of the kernel better.

Reply:
Yeah, much nicer.